src/derived.rs

```diff
@@ -44,6 +50,10 @@ where
     DB: Database,
     MP: MemoizationPolicy<DB, Q>,
 {
+    lru_cap: AtomicUsize,
+    // if `lru_keys` and `map` are locked togeter,
+    // `lru_keys` is locked first, to prevent deadlocks.
+    lru_keys: Mutex<LinkedHashSet<Q::Key>>,
     map: RwLock<FxHashMap<Q::Key, QueryState<DB, Q>>>,
```

Comment:
The alternative here, instead of "lock lru first", is to group both collections under the same lock:

```
map: RwLock<(FxHashMap<Q::Key, QueryState<DB, Q>>, LinkedHashSet<Q::Key>)>
```

This won't be as ergonomic, because, we'd have to map lock guards go get access to `FxHashMap` only. 

OTOH, now that I think about it, in the steady sate the LRU is full, so we have to lock both lru and map anyway, and collapsing that to just one lock seems like a win.